### PR TITLE
Add logstash system index APIs

### PR DESCRIPTION
**Pull request by jaymode**
This commit adds REST apis for logstash and its system index, `.logstash`, which is used to store pipelines for central management. The module adds new endpoints that enable CRUD operations
on the pipelines stored in the index.

### CODE REVIEW

File: x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/DeletePipelineAction.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.logstash.action;
+
+import org.elasticsearch.action.ActionType;
+
+public class DeletePipelineAction extends ActionType<org.elasticsearch.xpack.logstash.action.DeletePipelineResponse> {
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/DeletePipelineAction.java]: Aesthetic nitpick: Could we move this to an import so this is just `extends ActionType<DeletePipelineResponse>`?

File: x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/DeletePipelineResponse.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+package org.elasticsearch.xpack.logstash.action;
+
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+public class DeletePipelineResponse extends ActionResponse {
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/DeletePipelineResponse.java]: Why not make this an `AcknowledgedResponse`?

File: x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    @Inject
+    public TransportGetPipelineAction(TransportService transportService, ActionFilters actionFilters, Client client) {
+        super(GetPipelineAction.NAME, transportService, actionFilters, GetPipelineRequest::new);
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetPipelineRequest request, ActionListener<GetPipelineResponse> listener) {
+        if (request.ids().isEmpty()) {
+            client.prepareSearch(".logstash")
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java]: Given that `Logstash.LOGSTASH_CONCRETE_INDEX_NAME` is now public, why not use it here?

File: x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                                        e
+                                    )
+                                )
+                            );
+                        }
+                    };
+                    handleSearchResponse(searchResponse, pipelineSources, clearScroll, listener);
+                }, listener::onFailure));
+        } else if (request.ids().size() == 1) {
+            client.prepareGet(".logstash", request.ids().get(0)).setFetchSource(true).execute(ActionListener.wrap(response -> {
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java]: Same here about `Logstash.LOGSTASH_CONCRETE_INDEX_NAME`.

File: x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                if (response.isExists()) {
+                    listener.onResponse(new GetPipelineResponse(Map.of(response.getId(), response.getSourceAsBytesRef())));
+                } else {
+                    listener.onResponse(new GetPipelineResponse(Map.of()));
+                }
+            }, listener::onFailure));
+        } else {
+            client.prepareMultiGet()
+                .addIds(".logstash", request.ids())
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java]: Same here about `Logstash.LOGSTASH_CONCRETE_INDEX_NAME`.

File: x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                .execute(
+                    ActionListener.wrap(
+                        mGetResponse -> listener.onResponse(
+                            new GetPipelineResponse(
+                                Arrays.stream(mGetResponse.getResponses())
+                                    .filter(itemResponse -> itemResponse.isFailed() == false)
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java]: I'm not sure what to do instead, but if some of the GETs fail, do we really just want to silently swallow that? Should we at least log something here?

File: x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportPutPipelineAction.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Inject
+    public TransportPutPipelineAction(TransportService transportService, ActionFilters actionFilters, Client client) {
+        super(PutPipelineAction.NAME, transportService, actionFilters, PutPipelineRequest::new);
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(Task task, PutPipelineRequest request, ActionListener<PutPipelineResponse> listener) {
+        client.prepareIndex(".logstash")
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportPutPipelineAction.java]: Same here about `Logstash.LOGSTASH_CONCRETE_INDEX_NAME`.

File: x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/PipelineRequestResponseSerializationTests.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import org.elasticsearch.xpack.logstash.action.GetPipelineRequest;
+import org.elasticsearch.xpack.logstash.action.GetPipelineResponse;
+import org.elasticsearch.xpack.logstash.action.PutPipelineRequest;
+import org.elasticsearch.xpack.logstash.action.PutPipelineResponse;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class PipelineRequestResponseSerializationTests extends ESTestCase {
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/PipelineRequestResponseSerializationTests.java]: Weak recommendation: Break these out into `AbstractWireSerializingTestCase`s and get equals/hashcode tests for free, plus make it a bit easier to find the tests for each class.

**Review by AthenaEryma — COMMENTED** [review]
Looks good overall, left a few comments about tweaks I think we should make.

File: x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/DeletePipelineResponse.java (hunk 2 above)
Comment (jaymode) [reply x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/DeletePipelineResponse.java:15]: Is there a benefit? We're not acking anything here?

File: x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/PipelineRequestResponseSerializationTests.java (hunk 8 above)
Comment (williamrandolph) [reply x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/PipelineRequestResponseSerializationTests.java]: I've done this.

**Comment by williamrandolph** [comment]
@elasticmachine Please run elasticsearch-ci/2

File: x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/action/DeletePipelineRequestTests.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.logstash.action;
+
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.test.AbstractWireSerializingTestCase;
+
+public class DeletePipelineRequestTests extends AbstractWireSerializingTestCase<DeletePipelineRequest> {
```
Comment (gwbrown) [inline x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/action/DeletePipelineRequestTests.java:12]: Can you make these request/response tests implement `mutateInstance` as well? That's required to automatically check `.equals()` for the case of an object of the same type that's not equal.

**Review by AthenaEryma — APPROVED** [review]
Left a few very minor comments, but otherwise LGTM. No need for another round of review after addressing my comments here.

File: x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/action/DeletePipelineRequestTests.java (hunk 9 above)
Comment (williamrandolph) [reply x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/action/DeletePipelineRequestTests.java:12]: Done!

**Pull request merged**